Test scenarios

xsebek opened the pull request:
- add winning solutions to scenarios
- test the winning solutions
  - fail if any robot has a `Fatal error` in its log
  - timeout in X seconds (default 1)
- replace startup logging to `stdout` with marking errors
- fix creative mode ignoring scenario inventory
- fix integration test discovery (challenges are in a subfolder, so recurse)
- add a *Testing* scenario collection hidden without the `--cheat` flag
- closes #388

xsebek commented:
@byorgey I wrote a test case for #394, but then I needed `expectFail`, so I put the automatic testing of all scenarios to rest.

I also allow `testSolution'` to check the final `GameState` so automating that is probably not that useful.

xsebek commented:
I really like [tasty-expected-failure](https://hackage.haskell.org/package/tasty-expected-failure-0.12.3/docs/Test-Tasty-ExpectedFailure.html):
|![image](https://user-images.githubusercontent.com/44544735/173920035-8bf18fad-dd6a-4068-ade6-4f8746b05c3b.png)|
|-|

xsebek commented:
Looking at the image above, I noticed we do not actually test all scenarios. :sweat_smile: 

Now it is fixed. :heavy_check_mark:

byorgey commented on src/Swarm/Game/State.hs:
```diff
@@ -515,7 +522,7 @@ playScenario em scenario userSeed toRun g = do
         %~ case scenario ^. scenarioCreative of
           False -> id
-          True -> const (fromElems (map (0,) things))
+          True -> union (fromElems (map (0,) things))
```
What is the purpose of this change?

xsebek replied:
[Fix creative mode ignoring scenario inventory](https://github.com/swarm-game/swarm/pull/400/commits/602cefa70739883a076109495df622b17cbb9383) :slightly_smiling_face:

If I add inventory to the base in a scenario, originally it would overwrite it with zero of everything.
I could workaround by adding `program: "create X; create Y; ..."`, but I thought it was a bug, so I fixed it.

xsebek replied:
The reason why I looked for it was that the new test case would not work without inventory in creative mode. 

(I was also trying modified versions of it where you start with a logger and give it to the robot.)

I expect being able to specify inventory even in creative mode will come in handy. :slightly_smiling_face:

byorgey replied:
Aha, I see now! 👍

byorgey commented on src/Swarm/TUI/Controller.hs:
```diff
@@ -145,7 +145,7 @@ handleNewGameMenuEvent scenarioStack@(curMenu :| rest) s = \case
       Just (SICollection _ c) ->
         continue $
-          s & uiState . uiMenu .~ NewGameMenu (NE.cons (mkScenarioList c) scenarioStack)
+          s & uiState . uiMenu .~ NewGameMenu (NE.cons (mkScenarioList (s ^. gameState . creativeMode) c) scenarioStack)
```
I don't see how the `creativeMode` flag in the `gameState` is relevant here.  It doesn't really tell us much unless we're currently playing a game.  In particular it does *not* tell us whether Swarm was started with the `--cheat` flag enabled.

xsebek replied:
Well spotted! :+1: Weird that I did not notice, but I guess I did not test it too much. :sweat_smile:

xsebek replied:
Now it is `(s ^. uiState . uiCheatMode)` as it should have been. :heavy_check_mark:

byorgey commented on test/integration/Main.hs:
```diff
@@ -0,0 +1,162 @@
+  hasExt path = takeExtension path == ("." ++ ext)
+
+data Time
+  = -- | One second should be anough to run most programs.
```
```suggestion
  = -- | One second should be enough to run most programs.
```

xsebek replied:
Thanks, fixed :+1:

xsebek commented:
I rebased, cleaned up the history and added the new portal challenge solution to the tests :innocent: 

@byorgey given that there are some changes outside tests, I would like to merge the commits as they are now.
Unless you have more suggestions of course :slightly_smiling_face:

byorgey reviewed (approved):
Looks good!
